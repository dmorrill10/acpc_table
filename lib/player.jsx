import React from 'react';
import {CardList, BoardCards, PrivateCards} from './card_list';
import ChipStack from './chip_stack';

class Player extends React.Component {
  heldPlayerInformation() {
    const l = [(
        <div key='name' className='name'>{this.props.name}</div>
      ), (<CardList key='card-list' cards={this.props['hole-cards']}/>), (
        <div key='player_stack' className='player_stack'>
          <ChipStack key='chip_stack' amount={this.props['chip-stack-amount']}/>
        </div>
      )];
    return (
      <li key='held_player_information' className='held_player_information'>
        {l}
      </li>
    );
  }
  outerPlayerInformation() {
    let l = []
    if (this.props.dealer) {
      l.push(
        <div key='dealer_button' className='dealer_button'>Dealer</div>
      );
    }
    if (this.props.winnings && this.props.winnings > 0) {
      l.push((
        <div key='winnings' className='winnings'>
          <ChipStack amount={this.props.winnings}/>
        </div>
      ))
    } else if (this.props.contribution && this.props.contribution > 0) {
      l.push(
        <div key='contribution' className='contribution'>
          <ChipStack amount={this.props.contribution}/>
        </div>
      )
    } else {
      l.push(
        <div key='chip-stack-placeholder' className='chip-stack-placeholder'></div>
      );
    }
    return (
      <li key='outer_player_information' className='outer_player_information'>
        {l}
      </li>
    );
  }
  playerContents() {
    return [];
  }
  classes() {
    let c = 'player';
    if (this.props.acting) {
      c += ' acting';
    }
    if (this.props['has-folded'] || !this.props['hole-cards'] || this.props['hole-cards'].length < 1) {
      c += ' has-folded';
    }
    return c;
  }
  render() {
    return <ol className={this.classes()}>{this.playerContents()}</ol>;
  }
}
class User extends Player {
  classes() {
    return super.classes() + ' user';
  }
  playerContents() {
    return [this.outerPlayerInformation(), this.heldPlayerInformation()];
  }
}
class Opponent extends Player {
  playerContents() {
    return [this.heldPlayerInformation(), this.outerPlayerInformation()];
  }
}
export {Player, User, Opponent}
