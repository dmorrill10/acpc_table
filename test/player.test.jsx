import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {Player, User, Opponent} from '../dist/player.js';

// TODO Not sure how to make this test less brittle and unwieldy

describe('User', () => {
  describe('not next to act', () => {
    const isNextToAct = false;
    describe('without the dealer button', () => {
      const isDealer = false;
      describe('with a hand of two face-down cards', () => {
        const cards = [{}, {}];
        describe('receiving no chips', () => {
          const winnings = 0;
          describe('betting no chips', () => {
            const contribution = 0;
            describe('holding no chips', () => {
              const chipStackAmount = 0;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(false);
                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
            describe('holding chips', () => {
              const chipStackAmount = 9001;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(false);

                const heldPlayerInfo = wrapper.find('.player .held_player_information');
                expect(heldPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(9001);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
          });
          describe('betting chips', () => {
            const contribution = 100;
            describe('holding no chips', () => {
              const chipStackAmount = 0;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(false);

                const outerPlayerInfo = wrapper.find('.player .outer_player_information');
                expect(outerPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(100);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
            describe('holding chips', () => {
              const chipStackAmount = 9001;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(false);

                const heldPlayerInfo = wrapper.find('.player .held_player_information');
                expect(heldPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(9001);

                const outerPlayerInfo = wrapper.find('.player .outer_player_information');
                expect(outerPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(100);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
          });
        });
      });
      describe('with an implicitly folded hand', () => {
        const cards = [];
        describe('receiving no chips', () => {
          const winnings = 0;
          describe('betting no chips', () => {
            const contribution = 0;
            describe('holding no chips', () => {
              const chipStackAmount = 0;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(true);
                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
            describe('holding chips', () => {
              const chipStackAmount = 9001;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(true);

                const heldPlayerInfo = wrapper.find('.player .held_player_information');
                expect(heldPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(9001);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
          });
          describe('betting chips', () => {
            const contribution = 100;
            describe('holding no chips', () => {
              const chipStackAmount = 0;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(true);

                const outerPlayerInfo = wrapper.find('.player .outer_player_information');
                expect(outerPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(100);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
            describe('holding chips', () => {
              const chipStackAmount = 9001;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(true);

                const heldPlayerInfo = wrapper.find('.player .held_player_information');
                expect(heldPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(9001);

                const outerPlayerInfo = wrapper.find('.player .outer_player_information');
                expect(outerPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(100);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
          });
        });
      });
      describe('with an explicitly folded hand and two face-up cards', () => {
        const cards = [
          {
            rank: 'A',
            suit: 'c'
          }, {
            rank: 'K',
            suit: 'd'
          }
        ];
        const hasFolded = true;
        describe('receiving no chips', () => {
          const winnings = 0;
          describe('betting no chips', () => {
            const contribution = 0;
            describe('holding no chips', () => {
              const chipStackAmount = 0;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}
                  has-folded={hasFolded}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(hasFolded);
                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
            describe('holding chips', () => {
              const chipStackAmount = 9001;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}
                  has-folded={hasFolded}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(hasFolded);

                const heldPlayerInfo = wrapper.find('.player .held_player_information');
                expect(heldPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(9001);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
          });
          describe('betting chips', () => {
            const contribution = 100;
            describe('holding no chips', () => {
              const chipStackAmount = 0;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}
                  has-folded={hasFolded}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(hasFolded);

                const outerPlayerInfo = wrapper.find('.player .outer_player_information');
                expect(outerPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(100);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
            describe('holding chips', () => {
              const chipStackAmount = 9001;
              it('works', () => {
                const patient = (<User
                  name='player1'
                  hole-cards={cards}
                  dealer={isDealer}
                  acting={isNextToAct}
                  chip-stack-amount={chipStackAmount}
                  winnings={winnings}
                  contribution={contribution}
                  has-folded={hasFolded}/>);
                const wrapper = shallow(patient);

                expect(wrapper.find('.name').text()).toEqual('player1');
                expect(wrapper.find('.player').hasClass('acting')).toEqual(false);
                expect(wrapper.find('.player').hasClass('has-folded')).toEqual(hasFolded);

                const heldPlayerInfo = wrapper.find('.player .held_player_information');
                expect(heldPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(9001);

                const outerPlayerInfo = wrapper.find('.player .outer_player_information');
                expect(outerPlayerInfo.find('ChipStack').first().prop('amount')).toEqual(100);

                expect(renderer.create(patient).toJSON()).toMatchSnapshot();
              });
            });
          });
        });
      });
    });
  });
});
