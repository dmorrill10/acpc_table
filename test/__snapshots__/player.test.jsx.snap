exports[`User not next to act without the dealer button with a hand of two face-down cards receiving no chips betting chips holding chips works 1`] = `
<ol
  className="player user">
  <li
    className="outer_player_information">
    <div
      className="contribution">
      <div
        className="chip-stack">
        <div
          className="amount">
          100
        </div>
      </div>
    </div>
  </li>
  <li
    className="held_player_information">
    <div
      className="name">
      player1
    </div>
    <ol
      className="card-list">
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
    </ol>
    <div
      className="player_stack">
      <div
        className="chip-stack">
        <div
          className="amount">
          9001
        </div>
      </div>
    </div>
  </li>
</ol>
`;

exports[`User not next to act without the dealer button with a hand of two face-down cards receiving no chips betting chips holding no chips works 1`] = `
<ol
  className="player user">
  <li
    className="outer_player_information">
    <div
      className="contribution">
      <div
        className="chip-stack">
        <div
          className="amount">
          100
        </div>
      </div>
    </div>
  </li>
  <li
    className="held_player_information">
    <div
      className="name">
      player1
    </div>
    <ol
      className="card-list">
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
    </ol>
    <div
      className="player_stack" />
  </li>
</ol>
`;

exports[`User not next to act without the dealer button with a hand of two face-down cards receiving no chips betting no chips holding chips works 1`] = `
<ol
  className="player user">
  <li
    className="outer_player_information">
    <div
      className="chip-stack-placeholder" />
  </li>
  <li
    className="held_player_information">
    <div
      className="name">
      player1
    </div>
    <ol
      className="card-list">
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
    </ol>
    <div
      className="player_stack">
      <div
        className="chip-stack">
        <div
          className="amount">
          9001
        </div>
      </div>
    </div>
  </li>
</ol>
`;

exports[`User not next to act without the dealer button with a hand of two face-down cards receiving no chips betting no chips holding no chips works 1`] = `
<ol
  className="player user">
  <li
    className="outer_player_information">
    <div
      className="chip-stack-placeholder" />
  </li>
  <li
    className="held_player_information">
    <div
      className="name">
      player1
    </div>
    <ol
      className="card-list">
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
      <li>
        <div
          className="card">
          <div
            className="logo" />
        </div>
      </li>
    </ol>
    <div
      className="player_stack" />
  </li>
</ol>
`;
